nifi-stateless/nifi-stateless-bootstrap/src/main/java/org/apache/nifi/stateless/bootstrap/StatelessBootstrap.java

```diff
@@ -181,58 +166,75 @@ private static BlockListClassLoader createExtensionRootClassLoader(final File na
+            return Collections.emptySet();
+        }
 
-            final String absolutePath = file.getAbsolutePath();
-            if (absolutePath.startsWith(javaLibPath)) {
-                continue;
-            }
+        final File javaHome = new File(javaHomeValue);
+        if (!javaHome.exists()) {
+            logger.warn("System property for java.home is {} but that directory does not exist so will not allow any classes explicitly from java.home in AllowListClassLoader", javaHomeValue);
+            return Collections.emptySet();
+        }
 
-            findClassNamesInJar(file, classesFound);
-            jarFilesFound.add(file.getName());
+        final File[] javaHomeFiles = javaHome.listFiles();
+        if (javaHomeFiles == null) {
+            logger.warn("System property for java.home is {} but that directory is not readable so will not allow any classes explicitly from java.home in AllowListClassLoader", javaHomeValue);
+            return Collections.emptySet();
         }
+
+        final Set<File> loadableFiles = new HashSet<>();
+        for (final File file : javaHomeFiles) {
+            findLoadableFiles(file, loadableFiles);
+        }
+
+        return loadableFiles;
     }
 
-    private static void findClassLoaderUrls(final ClassLoader classLoader, final Set<URL> urls) {
-        if (classLoader == null) {
+    private static void findLoadableFiles(final File file, final Set<File> loadable) {
+        if (file.isDirectory()) {
+            final File[] children = file.listFiles();
+            if (children == null) {
+                return;
+            }
+
+            for (final File child : children) {
+                findLoadableFiles(child, loadable);
+            }
+
             return;
         }
 
-        if (classLoader instanceof URLClassLoader) {
-            final URLClassLoader urlClassLoader = (URLClassLoader) classLoader;
-            urls.addAll(Arrays.asList(urlClassLoader.getURLs()));
+        final String filename = file.getName();
+        if (filename.endsWith(".class") || filename.endsWith(".jar") || filename.endsWith(".jmod")) {
+            loadable.add(file);
         }
+    }
 
-        // If the classLoader is the system class loader, we are done. We don't want to process the parent of
-        // the system class loader (which would be the Launcher$ExtClassLoader that contains the JDK/JRE classes, etc)
-        if (classLoader != ClassLoader.getSystemClassLoader()) {
-            findClassLoaderUrls(classLoader.getParent(), urls);
+    private static void findLoadableClasses(final File file, final Set<String> classNames) throws IOException {
+        final String filename = file.getName();
+        if (filename.endsWith(".jar")) {
+            findClassNamesInJar(file, classNames);
+        } else if (filename.endsWith(".jmod")) {
+            findClassesInJmod(file, classNames);
```

Comment:
`.class` files are collected in `findLoadableFiles()` (line 227) but they are not handled here. It did not cause any issues in my flows but there can be other use cases when those classes are needed.